frontend/public/components/secrets/create-secret.tsx

```diff
@@ -0,0 +1,155 @@
+import * as _ from 'lodash-es';
+import * as React from 'react';
+import { Helmet } from 'react-helmet';
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+
+import { k8sCreate, k8sUpdate, K8sResourceKind } from '../../module/k8s';
+import { ButtonBar, Firehose, history, kindObj, StatusBox } from '../utils';
+import { getActiveNamespace, formatNamespacedRouteForResource, UIActions } from '../../ui/ui-actions';
+import { SafetyFirst } from '../safety-first';
+import { WebHookSecretSubform } from './subforms';
+
+const determineSecretTypeAbstraction = (secret) => {
+  const { data } = secret;
+  if (_.has(data, 'WebHookSecretKey')) {
+    return 'webhook';
+  }
+  return 'generic';
+};
+
+class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretState_> {
+  constructor (props) {
+    super(props);
+    let existingData = _.pick(props.obj, ['metadata', 'data']);
+    const secret = _.defaultsDeep({}, props.fixed, existingData, {
+      apiVersion: 'v1',
+      data: {},
+      kind: 'Secret',
+      metadata: {
+        name: '',
+      },
+      type: 'Opaque',
+    });
+
+    this.state = {
+      secretType: this.props.secretType || determineSecretTypeAbstraction(secret),
+      secret: secret,
+      inProgress: false,
+      type: secret.type,
+      stringData: _.mapValues(secret.data, (v) => window.atob(v)),
+    };
+    this.onNameChanged = this.onNameChanged.bind(this);
+    this.save = this.save.bind(this);
+  }
+  onDataChanged (secretsData) {
+    let stringData = {...this.state.stringData};
+    stringData = secretsData;
+    this.setState({stringData});
+  }
+  onNameChanged (event) {
+    let secret = {...this.state.secret};
+    secret.metadata.name = event.target.value;
+    this.setState({secret});
+  }
+  save (e) {
+    e.preventDefault();
+    const { kind, metadata } = this.state.secret;
+    this.setState({ inProgress: true });
+    const newSecret = _.assign({}, this.state.secret, {stringData: this.state.stringData});
+    const ko = kindObj(kind);
+    (this.props.isCreate
+      ? k8sCreate(ko, newSecret)
+      : k8sUpdate(ko, newSecret, metadata.namespace, newSecret.metadata.name)
+    ).then(
+      () => {
+        this.setState({inProgress: false});
+        if (metadata.namespace) {
+          this.props.setActiveNamespace(metadata.namespace);
+        }
+        history.push(formatNamespacedRouteForResource('secrets'));
+      },
+      err => this.setState({error: err.message, inProgress: false})
+    );
+  }
+  render () {
+    const title = `${this.props.titleVerb} ${_.upperFirst(this.state.secretType)} Secret`;
+    const { saveButtonText } = this.props;
+
+    let explanation = 'Webhook secret allow you to authenticate a webhook trigger.';
+    let subform = <WebHookSecretSubform onChange={this.onDataChanged.bind(this)} stringData={this.state.stringData} />;
```

Comment:
It would be nice if the base edit component didn't need to know about the specific secret types. I wonder if we can't use a higher-order component here:

https://reactjs.org/docs/higher-order-components.html#use-hocs-for-cross-cutting-concerns